File: src/truco/cfr.py
"""
External-Sampling MCCFR with CFR+ updates and linear averaging.

Strategy and regret tables are keyed by (info_key, action_key). The CFR
solver is agnostic to the keying scheme — pass `info_key_fn` and
`action_key_fn` (raw or abstract) at construction.

Terminal value of a round:
  - If `value_table` is None: use signed stake as utility (Phase 1-2).
  - Else: V[next_score, next_lead] from the score-state DP (Phase 3+).
"""
from __future__ import annotations
from collections import defaultdict
from typing import Callable, Optional
import random

from .game import State, legal_actions, step, deal
from .infoset import info_key_raw, action_key_raw


class CFRSolver:
    def __init__(
        self,
        info_key_fn: Callable = info_key_raw,
        action_key_fn: Callable = action_key_raw,
        value_table: Optional[dict] = None,
        rng: Optional[random.Random] = None,
    ):
        self.info_key_fn = info_key_fn
        self.action_key_fn = action_key_fn
        self.value_table = value_table
        self.rng = rng if rng is not None else random.Random(0)

        # info_key -> {action_key: float}
        self.regrets: dict = defaultdict(dict)
        self.strategy_sum: dict = defaultdict(dict)
        self.iteration: int = 0

    # ---------- terminal value ----------

    def terminal_value(self, s: State, traverser: int) -> float:
        assert s.outcome is not None
        w, stake = s.outcome.winner, s.outcome.stake_awarded
        if self.value_table is None:
            # Phase 1-2: round value = stake, signed for the traverser.
            return float(stake) if w == traverser else -float(stake)
        # Phase 3+: lookup V at the resulting score state, from traverser's view.
        s0, s1 = s.score
        if w == 0:
            new_score = (s0 + stake, s1)
        else:
            new_score = (s0, s1 + stake)
        # After a round, the lead alternates.
        next_lead = 1 - s.first_trick_starter
        v_p0 = self.value_table.get((new_score, next_lead), 0.0)
        return v_p0 if traverser == 0 else -v_p0

    # ---------- regret matching + ----------

    def _strategy(self, info, action_keys: list) -> dict:
        regrets = self.regrets[info]
        pos = {ak: max(regrets.get(ak, 0.0), 0.0) for ak in action_keys}
        total = sum(pos.values())
        if total > 0:
            return {ak: pos[ak] / total for ak in action_keys}
        u = 1.0 / len(action_keys)
        return {ak: u for ak in action_keys}

    # ---------- traversal ----------

    def traverse(self, s: State, traverser: int) -> float:
        if s.is_terminal:
            return self.terminal_value(s, traverser)

        p = s.to_act
        legal = legal_actions(s)
        mr = s.manilha_rank

        # Group raw legal actions by their (possibly abstracted) action key.
        # Under suit abstraction, two raw plays may collapse to one abstract action.
        by_key: dict = {}
        for a in legal:
            ak = self.action_key_fn(a, mr)
            by_key.setdefault(ak, a)  # any representative works; pick first.

        action_keys = list(by_key.keys())
        info = self.info_key_fn(s, p)
        sigma = self._strategy(info, action_keys)

        if p == traverser:
            values: dict = {}
            for ak in action_keys:
                raw = by_key[ak]
                values[ak] = self.traverse(step(s, raw), traverser)
            node_value = sum(sigma[ak] * values[ak] for ak in action_keys)

            # Regret update (CFR+).
            reg = self.regrets[info]
            for ak in action_keys:
                new_r = reg.get(ak, 0.0) + (values[ak] - node_value)
                reg[ak] = new_r if new_r > 0.0 else 0.0

            # Strategy averaging (linear).
            w = float(self.iteration)
            ssum = self.strategy_sum[info]
            for ak in action_keys:
                ssum[ak] = ssum.get(ak, 0.0) + w * sigma[ak]

            return node_value

        # Opponent: sample one action.
        ak = self._sample_from(sigma)
        return self.traverse(step(s, by_key[ak]), traverser)

    def _sample_from(self, sigma: dict):
        keys = list(sigma.keys())
        weights = [sigma[k] for k in keys]
        return self.rng.choices(keys, weights=weights, k=1)[0]

    # ---------- top-level loop ----------

    def iterate(
        self,
        state_factory: Callable[[random.Random], State],
        n_iters: int,
        on_progress: Optional[Callable[[int], None]] = None,
    ):
        for _ in range(n_iters):
            self.iteration += 1
            for traverser in (0, 1):
                s = state_factory(self.rng)
                self.traverse(s, traverser)
            if on_progress is not None and self.iteration % 1000 == 0:
                on_progress(self.iteration)

    # ---------- strategy extraction ----------

    def average_strategy(self) -> dict:
        out = {}
        for info, ssum in self.strategy_sum.items():
            total = sum(ssum.values())
            if total > 0:
                out[info] = {ak: v / total for ak, v in ssum.items()}
            else:
                n = len(ssum)
                out[info] = {ak: 1.0 / n for ak in ssum}
        return out

File: truco-rs/src/cfr.rs
//! External-sampling MCCFR with CFR+ or DCFR updates.
//!
//! Tables keyed by canonical string info/action keys (see `keys.rs`). Strategy
//! averaging is linear (CFR+) or DCFR-gamma weighted.

use std::collections::HashMap;
use rand::Rng;

use crate::game::{Action, State, legal_actions, step};
use crate::keys::{info_key_abstract, action_key_abstract};

pub type InfoKey = String;
pub type ActionKey = String;
pub type StrategyTable = HashMap<InfoKey, HashMap<ActionKey, f64>>;

#[derive(Clone, Copy, Debug)]
pub enum Variant { CfrPlus, Dcfr }

pub struct CFRSolver<'a> {
    pub regrets: HashMap<InfoKey, HashMap<ActionKey, f64>>,
    pub strategy_sum: HashMap<InfoKey, HashMap<ActionKey, f64>>,
    pub value_table: Option<&'a HashMap<(u8, u8, u8), f64>>,
    pub iteration: u64,
    pub variant: Variant,
    pub dcfr_alpha: f64,
    pub dcfr_beta: f64,
    pub dcfr_gamma: f64,
}

impl<'a> CFRSolver<'a> {
    pub fn new(variant: Variant, value_table: Option<&'a HashMap<(u8, u8, u8), f64>>) -> Self {
        Self {
            regrets: HashMap::new(),
            strategy_sum: HashMap::new(),
            value_table,
            iteration: 0,
            variant,
            dcfr_alpha: 1.5,
            dcfr_beta: 0.0,
            dcfr_gamma: 2.0,
        }
    }

    fn terminal_value(&self, s: &State, traverser: u8) -> f64 {
        let o = s.outcome.expect("terminal needs outcome");
        let w = o.winner;
        let stake = o.stake_awarded;
        match self.value_table {
            None => if w == traverser { stake as f64 } else { -(stake as f64) },
            Some(vt) => {
                let new_score = if w == 0 {
                    (s.score.0 + stake, s.score.1)
                } else {
                    (s.score.0, s.score.1 + stake)
                };
                let next_lead = 1 - s.first_trick_starter;
                let v_p0 = if new_score.0 >= 12 {
                    1.0
                } else if new_score.1 >= 12 {
                    -1.0
                } else {
                    *vt.get(&(new_score.0, new_score.1, next_lead)).unwrap_or(&0.0)
                };
                if traverser == 0 { v_p0 } else { -v_p0 }
            }
        }
    }

    fn regret_match(&self, info: &str, n: usize, action_keys: &[ActionKey]) -> Vec<f64> {
        let mut pos = vec![0.0; n];
        let mut total = 0.0;
        if let Some(rs) = self.regrets.get(info) {
            for (i, ak) in action_keys.iter().enumerate() {
                let r = rs.get(ak).copied().unwrap_or(0.0).max(0.0);
                pos[i] = r;
                total += r;
            }
        }
        if total > 0.0 {
            for v in &mut pos { *v /= total; }
            pos
        } else {
            vec![1.0 / n as f64; n]
        }
    }

    pub fn traverse<R: Rng>(&mut self, s: &State, traverser: u8, rng: &mut R) -> f64 {
        if s.is_terminal {
            return self.terminal_value(s, traverser);
        }
        let p = s.to_act;
        let legal = legal_actions(s);
        let mr = s.manilha_rank;

        // Group by action key (suit-abstraction collapses some plays).
        let mut by_key: Vec<(ActionKey, Action)> = Vec::with_capacity(legal.len());
        for a in legal {
            let ak = action_key_abstract(a, mr);
            if by_key.iter().all(|(k, _)| *k != ak) {
                by_key.push((ak, a));
            }
        }
        let action_keys: Vec<ActionKey> = by_key.iter().map(|(k, _)| k.clone()).collect();
        let info = info_key_abstract(s, p);
        let sigma = self.regret_match(&info, action_keys.len(), &action_keys);

        if p == traverser {
            let mut values = vec![0.0f64; action_keys.len()];
            for (i, (_, a)) in by_key.iter().enumerate() {
                let ns = step(s, *a);
                values[i] = self.traverse(&ns, traverser, rng);
            }
            let node_value: f64 = sigma.iter().zip(values.iter()).map(|(s, v)| s * v).sum();

            let t = self.iteration as f64;
            let (pos_factor, neg_factor) = match self.variant {
                Variant::CfrPlus => (1.0, 0.0), // negatives clamped to 0
                Variant::Dcfr => {
                    let pf = t.powf(self.dcfr_alpha) / (t.powf(self.dcfr_alpha) + 1.0);
                    let nf = t.powf(self.dcfr_beta) / (t.powf(self.dcfr_beta) + 1.0);
                    (pf, nf)
                }
            };

            let reg_entry = self.regrets.entry(info.clone()).or_insert_with(HashMap::new);
            for (i, ak) in action_keys.iter().enumerate() {
                let r_cur = reg_entry.get(ak).copied().unwrap_or(0.0);
                let r_new = r_cur + (values[i] - node_value);
                let r_after = if r_new > 0.0 { r_new * pos_factor } else { r_new * neg_factor };
                reg_entry.insert(ak.clone(), r_after);
            }

            // Strategy averaging: linear for CFR+, gamma-weighted for DCFR.
            let weight = match self.variant {
                Variant::CfrPlus => t,
                Variant::Dcfr => {
                    let denom = t + 1.0;
                    (t / denom).powf(self.dcfr_gamma) * t
                }
            };
            let ssum = self.strategy_sum.entry(info).or_insert_with(HashMap::new);
            for (i, ak) in action_keys.iter().enumerate() {
                let v = ssum.get(ak).copied().unwrap_or(0.0);
                ssum.insert(ak.clone(), v + weight * sigma[i]);
            }

            node_value
        } else {
            // Sample opponent.
            let r: f64 = rng.gen();
            let mut acc = 0.0;
            let mut chosen = sigma.len() - 1;
            for (i, p) in sigma.iter().enumerate() {
                acc += p;
                if r <= acc { chosen = i; break; }
            }
            let (_, a) = by_key[chosen];
            let ns = step(s, a);
            self.traverse(&ns, traverser, rng)
        }
    }

    pub fn average_strategy(&self) -> StrategyTable {
        let mut out = HashMap::with_capacity(self.strategy_sum.len());
        for (info, ssum) in &self.strategy_sum {
            let total: f64 = ssum.values().sum();
            let mut dist = HashMap::with_capacity(ssum.len());
            if total > 0.0 {
                for (ak, v) in ssum {
                    dist.insert(ak.clone(), v / total);
                }
            } else {
                let n = ssum.len() as f64;
                for ak in ssum.keys() {
                    dist.insert(ak.clone(), 1.0 / n);
                }
            }
            out.insert(info.clone(), dist);
        }
        out
    }
}
